kimchi/src/circuits/expr.rs

```diff
@@ -540,8 +551,148 @@ pub enum Expr<C> {
+                                (c2_reduced, false)
+                            }
+                        } else if reduce_further2 && c2_reduced.is_one() {
+                            (c1_reduced, false)
+                        } else {
+                            (
+                                BinOp(Op2::Mul, Box::new(c1_reduced), Box::new(c2_reduced)),
+                                false,
+                            )
+                        }
+                    }
+                }
+            }
+            Pow(c, power) => {
+                let (c_reduced, reduce_further) = c.apply_feature_flags_inner(features);
+                if reduce_further && (c_reduced.is_zero() || c_reduced.is_one()) {
+                    (c_reduced, true)
+                } else {
+                    (Pow(Box::new(c_reduced), *power), false)
+                }
+            }
+            Cache(cache_id, c) => {
+                let (c_reduced, reduce_further) = c.apply_feature_flags_inner(features);
+                if reduce_further {
+                    (c_reduced, true)
+                } else {
+                    (Cache(*cache_id, Box::new(c_reduced)), false)
+                }
+            }
+            IfFeature(feature, c1, c2) => {
+                let is_enabled = {
+                    use FeatureFlag::*;
+                    match feature {
+                        ChaCha => features.chacha,
+                        RangeCheck => features.range_check,
+                        ForeignFieldAdd => features.foreign_field_add,
+                        ForeignFieldMul => features.foreign_field_mul,
+                        Xor => features.xor,
+                        Rot => features.rot,
+                        LookupTables => {
+                            features.lookup_features.patterns != LookupPatterns::default()
+                        }
+                        RuntimeLookupTables => features.lookup_features.uses_runtime_tables,
+                        LookupPattern(pattern) => features.lookup_features.patterns[*pattern],
+                        TableWidth(width) => features
+                            .lookup_features
+                            .patterns
+                            .into_iter()
+                            .any(|feature| feature.max_joint_size() >= (*width as u32)),
+                        LookupsPerRow(count) => features
+                            .lookup_features
+                            .patterns
+                            .into_iter()
+                            .any(|feature| feature.max_lookups_per_row() >= (*count as usize)),
+                    }
+                };
+                if is_enabled {
+                    let (c1_reduced, _) = c1.apply_feature_flags_inner(features);
+                    (c1_reduced, false)
+                } else {
+                    let (c2_reduced, _) = c2.apply_feature_flags_inner(features);
+                    (c2_reduced, true)
```

Comment:
Mmm interesting different behaviour for c1 and c2 here... what is the motivation for `is_enabled` to act like this?

Reply:
lookup stuff:

```rust
dummy_used = E::IfFeature(
                        FeatureFlag::LookupsPerRow((i + 1) as isize),
                        Box::new(dummy_used),
                        Box::new(E::one()),
                    );
```